shazar for configuration

dsugden opened the pull request:
For discussion...

huntc commented on sbt-bundle-tester/src/main/configuration/backend/bundle.conf:
```diff
@@ -0,0 +1,21 @@
+  "override-1.0.0" = {
+    description      = "override"
+    file-system-type = "universal"
+    start-command    = ["override-1.0.0/bin/override", "-J-Xms67108864", "-J-Xmx67108864"]
```
A nicer example of a configuration override would be to have to just:

```
components = {
  "override-1.0.0" = {
    start-command    = ["override-1.0.0/bin/override", "-J-Xms67108864", "-J-Xmx67108864"]
  }
}
```

i.e. just so what is being overloaded.

huntc commented on sbt-bundle-tester/src/main/configuration/frontend/bundle.conf:
```diff
@@ -0,0 +1,21 @@
+      }
+    }
+  }
+}
```
Same deal as for backend i.e. a smaller example illustrating that you don't have to type out all of the config.

huntc commented on src/main/scala/com/typesafe/sbt/bundle/SbtBundle.scala:
```diff
@@ -50,6 +50,11 @@ object Import {
+    val configurationPath = SettingKey[String](
+      "configuration-path",
+      "The Location of the additional configuration to use"
+    )
```
`configurationDir` may be more conventional. Also type it as a `File`.

huntc replied:
Comment outstanding.

huntc commented on src/main/scala/com/typesafe/sbt/bundle/SbtBundle.scala:
```diff
@@ -105,6 +110,8 @@ object Import {
 
   val Bundle = config("bundle") extend Universal
+
+  val Configuration = config("config")
```
I think that this should also extend the `Universal` configuration.

dsugden replied:
Sure. I took it out because I noticed it was doing a lot of unnecessary stuff, ie. creating POMs etc.  ?

huntc commented on src/main/scala/com/typesafe/sbt/bundle/SbtBundle.scala:
```diff
@@ -123,6 +130,8 @@ object SbtBundle extends AutoPlugin {
   override def trigger = AllRequirements
 
+  private val tmpConfName = "frontend"
+
```
WAT?

dsugden replied:
this is a temp hack... :)

dsugden replied:
Therein lies my confusion:  how to specify which dir's to shazar...

huntc commented on src/main/scala/com/typesafe/sbt/bundle/SbtBundle.scala:
```diff
@@ -145,8 +154,21 @@ object SbtBundle extends AutoPlugin {
+    NativePackagerKeys.stagingDirectory in Configuration := (target in Configuration).value / "stage",
+    target in Bundle := target.value / "bundle",
+    target in Configuration := target.value / "configuration",
+    configurationPath := baseDirectory.value.getAbsolutePath + "/src/main/configuration/"
```
This should absolutely be expressed as a path in order to remain platform independent i.e.

```scala
configurationPath := sourceDirectory.value / "configuration"
```

The above also uses the default source directory and is consistent with the way native packager picks up templates etc. from src i.e. there is no distinction between "main" and "test" - there doesn't need to be as this is for deployment...

huntc replied:
In fact, instead of `configurationPath`:

```scala
sourceDirectory in Configuration := sourceDirectory.value / "configuration"
```

huntc commented on src/main/scala/com/typesafe/sbt/bundle/SbtBundle.scala:
```diff
@@ -169,6 +191,37 @@ object SbtBundle extends AutoPlugin {
+   */
+  private def createConfiguration(): Def.Initialize[Task[File]] = Def.task {
+
+    val name = "frontend"
```
We should have a string setting that specifies the configuration to use. Perhaps the default for this setting can simply be "default". How about `configurationName`?

dsugden replied:
Perfect!

dsugden commented:
I've addressed the comments here. I changed Configuration to BundleConfiguration as "Configuration" risks name clashes in scripted test build.sbt.  

Added configurationName as a Setting. It defaults to "default". If there is no directory matching the configurationName setting:
    sys.error(s"Directory $srcDir does not exist. Specify the configuration directory with the 'configurationName' setting")

huntc commented on src/main/scala/com/typesafe/sbt/bundle/SbtBundle.scala:
```diff
@@ -151,8 +163,22 @@ object SbtBundle extends AutoPlugin {
-    target in Bundle := target.value / "bundle"
+    NativePackagerKeys.stagingDirectory in BundleConfiguration := (target in BundleConfiguration).value / "stage",
+    target in Bundle := target.value / "bundle",
+    target in BundleConfiguration := target.value / "configuration",
```
Perhaps `bundle-configuration` then to be consistent?

huntc commented on src/main/scala/com/typesafe/sbt/bundle/SbtBundle.scala:
```diff
@@ -175,6 +201,24 @@ object SbtBundle extends AutoPlugin {
+    IO.move(archive, hashArchive)
+    streams.value.log.info(s"Configuration Bundle has been created: $hashArchive")
+    hashArchive
+  }
```
I'm now wondering if some of the above function can be generalised and re-used from `createBundle`? WDYT?

huntc replied:
A lot of the above is doing what `shazar` does in case you're looking for inspiration when it comes to names.

huntc commented on src/main/scala/com/typesafe/sbt/bundle/SbtBundle.scala:
```diff
@@ -248,6 +292,15 @@ object SbtBundle extends AutoPlugin {
+    IO.copyDirectory(srcDir, configurationTarget, true, true)
+    configurationTarget
+  }
+
```
Ditto in terms of generalisation.

huntc commented on src/sbt-test/sbt-bundle/basic/build.sbt:
```diff
@@ -53,3 +55,15 @@ checkBundleConf := {
   contents should include(expectedContents)
 }
+
+val checkConfigDist = taskKey[Unit]("check-config-dist-contents")
```
We should have a separate scripted test for testing config - perhaps leave the existing test as-is.

huntc commented:
Looking really nice! (a few comments, and please squash commits)
